### PR TITLE
fix: Oracle dispatch error & minor macro refactor

### PR DESCRIPTION
**Pull request by KirilMihaylov**
Addition: Minor refactor to `oracle` macro to add the base currency as a parameter.

### CODE REVIEW

File: packages/marketprice/src/alarms/mod.rs [hunk 1]
```diff
@@ -395,6 +395,68 @@ pub mod tests {
...
     }
 
+    #[test]
+    fn test_out_for_delivery_removes_above() {
+        const PRICE_BASE: Amount = 1;
+        const PRICE_QUOTE: Amount = if let Some(quote) = PRICE_BASE.checked_add(1) {
+            quote
+        } else {
+            panic!()
+        };
```
Comment (gmanev7) [inline packages/marketprice/src/alarms/mod.rs:405]: [everywhere] It's an overkill ... please use the operator + and - ... after all we always panic when overflow

File: contracts/oracle/src/contract/oracle/mod.rs [hunk 2]
```diff
@@ -155,3 +155,214 @@ where
...
+    #[track_caller]
+    fn add_higher_alarm(storage: &mut dyn Storage) {
+        add_alarm(storage, 1, 3)
+    }
+
+    #[track_caller]
+    fn add_lower_alarm(storage: &mut dyn Storage) {
+        add_alarm(storage, 2, 2)
+    }
+
```
Comment (gmanev7) [inline contracts/oracle/src/contract/oracle/mod.rs:297]: replace this code with price constants and pass them to `add_alarm` not `Amount`-s

File: contracts/oracle/src/contract/oracle/mod.rs [hunk 3]
```diff
@@ -155,3 +155,214 @@ where
...
+    #[track_caller]
+    fn feed_below_price(price_config: &PriceConfig, storage: &mut dyn Storage, now_seconds: u64) {
+        feed_price(price_config, storage, now_seconds, 1, 1)
+    }
+
+    #[track_caller]
+    fn feed_normal_price(price_config: &PriceConfig, storage: &mut dyn Storage, now_seconds: u64) {
+        feed_price(price_config, storage, now_seconds, 1, 2)
+    }
```
Comment (gmanev7) [inline contracts/oracle/src/contract/oracle/mod.rs:326]: get rid of this chunk simmilarly to the `add_alarm` comment above

File: contracts/oracle/src/contract/oracle/mod.rs [hunk 4]
```diff
@@ -155,3 +155,214 @@ where
...
+        now_seconds += 1;
+
+        // Bug happens on this step.
+        dispatch_0(&mut storage, now_seconds);
+
+        feed_below_price(&price_config, &mut storage, now_seconds);
+
+        now_seconds += 1;
+
+        dispatch_0(&mut storage, now_seconds);
```
Comment (gmanev7) [inline contracts/oracle/src/contract/oracle/mod.rs:241]: please structure it further ... now it is ~ 50 lines long and is difficult to read

File: packages/marketprice/src/alarms/mod.rs [hunk 5]
```diff
@@ -395,6 +395,68 @@ pub mod tests {
...
+
+        alarms.last_delivered().unwrap();
+
+        alarms.ensure_no_in_delivery().unwrap();
+
+        // Catch below
+        assert_eq!(alarms.alarms(lower_price).count(), 0);
+
+        // Catch above or equal
+        assert_eq!(alarms.alarms(price).count(), 0);
```
Comment (gmanev7) [inline packages/marketprice/src/alarms/mod.rs:457]: please structure it further ... now it is ~ 60 lines long and is difficult to read

**Review by gmanev7 — COMMENTED** [review]
LGTM except how easy it is to read some long functions ... and a few constants spread here and there

File: contracts/oracle/src/contract/oracle/mod.rs [hunk 6]
```diff
@@ -155,3 +155,137 @@ where
...
+            Oracle::<'_, _, BaseCurrency>::load(storage).unwrap();
+
+        let alarms: u32 = oracle.try_notify_alarms(NOW, 16).unwrap().0;
+
+        assert_eq!(alarms, expected_count);
+    }
+
+    #[track_caller]
+    fn deliver(storage: &mut dyn Storage, count: u32) {
+        let storage: &mut dyn Storage = storage;
```
Comment (gmanev7) [inline contracts/oracle/src/contract/oracle/mod.rs:276]: ???

Comment (KirilMihaylov) [reply contracts/oracle/src/contract/oracle/mod.rs:276]: Leftover from the minimization. Removing it now.

File: packages/finance/src/price/mod.rs [hunk 7]
```diff
@@ -66,23 +66,45 @@ where
...
     }
 
+    pub const fn base_amount(&self) -> Coin<C> {
+        self.amount
+    }
+
+    pub const fn quote_amount(&self) -> Coin<QuoteC> {
+        self.amount_quote
+    }
+
```
Comment (gmanev7) [inline packages/finance/src/price/mod.rs:107]: Price is not a bag of two amounts everyone should know about. please remove. Moreover you use them only for tests

File: packages/finance/src/coin/mod.rs [hunk 8]
```diff
@@ -77,16 +80,21 @@ where
...
 
-        debug_assert_eq!(self.amount % gcd, 0);
-        debug_assert_eq!(other.amount % gcd, 0);
+        #[cfg(debug_assertions)]
+        if self.amount % gcd != 0 {
+            panic!("LHS-value's amount is not divisible by the GCD!");
+        } else if other.amount % gcd != 0 {
+            panic!("RHS-value's amount is not divisible by the GCD!");
+        }
+
```
Comment (gmanev7) [inline packages/finance/src/coin/mod.rs:97]: a simple `debug_assert!()`  would suffice!

File: packages/finance/src/coin/mod.rs [hunk 9]
```diff
@@ -40,10 +39,14 @@ where
...
 
-    pub fn is_zero(&self) -> bool {
+    pub const fn is_zero(&self) -> bool {
         self.amount == Zero::ZERO
     }
 
+    pub const fn into_amount(self) -> Amount {
+        self.amount
+    }
+
```
Comment (gmanev7) [inline packages/finance/src/coin/mod.rs:49]: please remove it, I understand the motivation ... but that should not happen at this price

File: packages/finance/src/coin/mod.rs (hunk 8 above)
Comment (KirilMihaylov) [reply packages/finance/src/coin/mod.rs:97]: With all my respect, it does not. I am adding `const` modifier which doesn't play nice with `{|debug_}assert{|_eq|_ne}!` macros. If I didn't have to change it, I wouldn't have.

File: packages/finance/src/coin/mod.rs (hunk 9 above)
Comment (KirilMihaylov) [reply packages/finance/src/coin/mod.rs:49]: We don't have a way to neither do arithmetic, nor use trait methods in constant time. What do you suggest that I do?

File: packages/finance/src/price/mod.rs (hunk 7 above)
Comment (KirilMihaylov) [reply packages/finance/src/price/mod.rs:107]: That would reintroduce more constants in the tests again, but I'll do as you wish.

File: packages/finance/src/price/mod.rs [hunk 10]
```diff
@@ -66,23 +66,45 @@ where
...
+        #[cfg(debug_assertions)]
+        res.invariant_held().unwrap();
+
+        res
+    }
+
+    #[cfg(feature = "testing")]
+    pub const fn unchecked(amount: Coin<C>, amount_quote: Coin<QuoteC>) -> Self {
+        Self::new_inner(amount, amount_quote)
+    }
```
Comment (gmanev7) [inline packages/finance/src/price/mod.rs:80]: there is almost the same alternative, why have you added this?

Comment (KirilMihaylov) [reply packages/finance/src/price/mod.rs:80]: `invariant_held` can not be ran in compile time without further refactoring, thus the need for this method.

File: packages/finance/src/coin/mod.rs (hunk 8 above)
Comment (gmanev7) [reply packages/finance/src/coin/mod.rs:93]: a local compilation passes with not errors or warnings

File: packages/finance/src/coin/mod.rs [hunk 11]
```diff
@@ -77,16 +76,26 @@ where
...
 
-        debug_assert_eq!(self.amount % gcd, 0);
-        debug_assert_eq!(other.amount % gcd, 0);
+        debug_assert!(
+            self.amount % gcd == 0,
+            "LHS-value's amount is not divisible by the GCD!"
+        );
+        debug_assert!(
+            other.amount % gcd == 0,
+            "RHS-value's amount is not divisible by the GCD!"
```
Comment (gmanev7) [inline packages/finance/src/coin/mod.rs:96]: the description should be inverted

**Review by gmanev7 — APPROVED** [review]

**Pull request merged**